File: 2958-length-of-longest-subarray-with-at-most-k-frequency/2958-length-of-longest-subarray-with-at-most-k-frequency.cpp
class Solution {
public:
    int maxSubarrayLength(vector<int>& nums, int k) {
        unordered_map<int,int>freq;
        int res=0,n=nums.size();
        int tail=0,head=-1;
        while(head<n)
        {
            while(head+1<n&&(!freq.count(nums[head+1])||freq[nums[head+1]]+1<=k))
            {
                head++;
                freq[nums[head]]++;
            }
            res=max(res,head-tail+1);
            if(tail>head)
            {
                tail++;
                head=tail-1;
            }
            else 
            {
                freq[nums[tail]]--;
                tail++;
            }
        }
        return res;
    }
};

File: 2958-length-of-longest-subarray-with-at-most-k-frequency/2958-length-of-longest-subarray-with-at-most-k-frequency.java
class Solution {
    public int maxSubarrayLength(int[] nums, int k) {
        Map<Integer,Integer> mp = new HashMap<>();
        int mx=1,left=-1,curr=0;
        while(curr<nums.length)
        {
            while(mp.getOrDefault(nums[curr],0)==k)
            {
                left++;
                mp.put(nums[left],mp.get(nums[left])-1);
            }
            mp.put(nums[curr],mp.getOrDefault(nums[curr],0)+1);
            mx=Math.max(mx,curr-left);
            curr++;
        }
        return mx;
    }
}